### PR TITLE
Project scoped serializers

### PR DESCRIPTION
**Pull request by yakkomajuri**
## Changes

I started working on converting the broken plugins that use the PostHog API to the new `posthog.api` util, and then realized that while we made all the viewsets accept a team specification before (https://github.com/PostHog/posthog/pull/5044), the serializers weren't covered, meaning passing the token didn't work for requests like`POST` or `PATCH` to the models.

I then started to work on this only to realize (while writing tests) that we just recently moved (started moving) to project-based endpoints anyway (#6430). So now I'm wondering what to do here. 

This seems like a change to what is soon to become legacy code, but also needed if we want to fix the broken plugins. How far along are we with a full rollout of the project-based API changes?

Do we look to get this in, or should I start updating the `posthog.api` util in the plugin server to the new approach?

Stopped my work here to ask for directions, and putting this PR out just for context.

 @Twixes would love your thoughts before doing more work here

## How did you test this code?

Ran all API tests for compatibility, added test.

### CODE REVIEW

**Comment by yakkomajuri** [comment]
@Twixes care leaving your thoughts from Slack here for context completeness?

**Comment by Twixes** [comment]
Sure, we had a quick chat with Yakko about this.
So we can still use `get_token` just like before, the refactors have been done in such a way that there is no change for project API token handling for old-style endpoints (non-project-nested). The problem is that just _can't_ work for the new-style endpoints. And we will stick to those going forward – old ones will be removed. Maybe the `get_token` replacement for project-nesting would be to just allow both ID and the API token for specifying the project like so `/api/projects/<id_or_api_token>/`.

**Comment by yakkomajuri** [comment]
This should be good to ~~go~~ be reviewed @Twixes 

`StructuredViewsetMixin` already handles token being included from previous changes.

File: posthog/api/person.py [hunk 1]
```diff
@@ -11,7 +11,7 @@
 from rest_framework.utils.serializer_helpers import ReturnDict
 from rest_framework_csv import renderers as csvrenderers
 
-from posthog.api.routing import StructuredViewSetMixin
+from posthog.api.routing import ProjectScopedHyperlinkedModelSerializer, StructuredViewSetMixin
```
Comment (Twixes) [inline posthog/api/person.py:14]: Import seems unused?

File: posthog/api/event.py [hunk 2]
```diff
@@ -16,7 +16,7 @@
 from rest_framework.settings import api_settings
 from rest_framework_csv import renderers as csvrenderers
 
-from posthog.api.routing import StructuredViewSetMixin
+from posthog.api.routing import ProjectScopedModelSerializer, StructuredViewSetMixin
```
Comment (Twixes) [inline posthog/api/event.py:19]: Import seems unused?

File: posthog/api/test/test_feature_flag.py [hunk 3]
```diff
@@ -113,6 +114,19 @@ def test_create_feature_flag(self, mock_capture):
...
+        self.client.force_login(self.user)
+
+        response = self.client.post(
+            f"/api/feature_flag/?token={team2.api_token}",
+            {"name": "Alpha feature", "key": "alpha-feature", "filters": {"groups": [{"rollout_percentage": 50}]}},
+            format="json",
+        )
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        instance = FeatureFlag.objects.get(id=response.json()["id"])
+        self.assertEqual(instance.team, team2)
```
Comment (Twixes) [inline posthog/api/test/test_feature_flag.py:128]: Do we have some tests for the non-legacy way?

Comment (yakkomajuri) [reply posthog/api/test/test_feature_flag.py:128]: Well I didn't touch the new stuff and there were tests for it before. Can add an extra one though

Comment (yakkomajuri) [reply posthog/api/test/test_feature_flag.py:128]: Which should override? We currently have it in `master` such that a token passed in as data will override the URL one, which is probably incorrect?

Comment (Twixes) [reply posthog/api/test/test_feature_flag.py:128]: Basically I mean a test that'd be a working example of the project-nesting-compatible way of using `token` in the API (unless there already is one and I just missed it). Regarding overrides, hm, IMO the API should be structured in a way that makes it impossible to specify both `id` and `api_token`, as that can only result in a confusing conflict.

**Comment by Twixes** [comment]
Do we still want to get this in @yakkomajuri? I can re-review if the merge conflicts are resolved (not sure if the last point about a test was resolved)

**Comment by yakkomajuri** [comment]
The reason I halted here is that currently in production the query project token overrides the path project id. And given that that's not desirable, I felt I should fix that in the work I do here. And then I ignored it :D

**Comment by yakkomajuri** [comment]
I do still want this in. Will revisit next week

**Pull request closed without merging**